kotlinx-coroutines-core/common/src/flow/operators/Zip.kt

```diff
@@ -209,46 +297,22 @@ private fun CoroutineScope.asFairChannel(flow: Flow<*>): ReceiveChannel<Any> = p
  * }
  * ```
  */
+@JvmName("flowZip")
 @ExperimentalCoroutinesApi
-public fun <T1, T2, R> Flow<T1>.zip(other: Flow<T2>, transform: suspend (T1, T2) -> R): Flow<R> = flow {
-    coroutineScope {
-        val first = asChannel(this@zip)
-        val second = asChannel(other)
-        /*
-         * This approach only works with rendezvous channel and is required to enforce correctness
-         * in the following scenario:
-         * ```
-         * val f1 = flow { emit(1); delay(Long.MAX_VALUE) }
-         * val f2 = flowOf(1)
-         * f1.zip(f2) { ... }
-         * ```
-         *
-         * Invariant: this clause is invoked only when all elements from the channel were processed (=> rendezvous restriction).
-         */
-        (second as SendChannel<*>).invokeOnClose {
-            if (!first.isClosedForReceive) first.cancel(AbortFlowException())
-        }
-
-        val otherIterator = second.iterator()
-        try {
-            first.consumeEach { value ->
-                if (!otherIterator.hasNext()) {
-                    return@consumeEach
-                }
-                val secondValue = NULL.unbox<T2>(otherIterator.next())
-                emit(transform(NULL.unbox(value), NULL.unbox(secondValue)))
-            }
-        } catch (e: AbortFlowException) {
-            // complete
-        } finally {
-            if (!second.isClosedForReceive) second.cancel(AbortFlowException())
-        }
-    }
-}
+public fun <T1, T2, R> Flow<T1>.zip(other: Flow<T2>, transform: suspend (T1, T2) -> R): Flow<R> = zipImpl(this, other, transform)
```

Comment:
I'd suggest to get rid of top-level `zip`. Its name is too short to have it a top level function and, unlike, `combine`, there is no "consistency" argument for having it at the top level.

Reply:
Good point